ompi/mca/coll/shared/coll_shared_reduce.c

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2019      The University of Tennessee and The University
+ *                         of Tennessee Research Foundation.  All rights
+ *                         reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "coll_shared.h"
+
+int mca_coll_shared_reduce_intra(const void *sbuf, void *rbuf, int count,
+                                 struct ompi_datatype_t *dtype,
+                                 struct ompi_op_t *op,
+                                 int root,
+                                 struct ompi_communicator_t *comm,
+                                 mca_coll_base_module_t * module)
+{
+    ptrdiff_t extent, lower_bound;
+
+    ompi_datatype_get_extent(dtype, &lower_bound, &extent);
+    if (count * extent <= 32 * 1024) {
+        ompi_coll_base_reduce_intra_binomial(sbuf, rbuf, count, dtype, op,
+                                             root, comm, module, 2048,
+                                             128);
+    } else {
+        mca_coll_shared_reduce_ring_intra(sbuf, rbuf, count, dtype, op,
+                                          root, comm, module);
+    }
+
+    return OMPI_SUCCESS;
+
+}
+
+/**
+ * Each process operates a part of the shared data buffer in turn.
+ * Suppose the number of processes is 4.
+ * Step 1:
+ * |  P0  |  P1  |  P2  |  P3  |
+ * Step 2:
+ * |  P1  |  P2  |  P3  |  P0  |
+ * Step 3:
+ * |  P2  |  P3  |  P0  |  P1  |
+ * Step 4:
+ * |  P3  |  P0  |  P1  |  P2  |
+ * At last, root copies data back from the shared data buffer.
+ */
+int mca_coll_shared_reduce_ring_intra(const void *sbuf, void *rbuf,
+                                      int count,
+                                      struct ompi_datatype_t *dtype,
+                                      struct ompi_op_t *op, int root,
+                                      struct ompi_communicator_t *comm,
+                                      mca_coll_base_module_t * module)
+{
+    mca_coll_shared_module_t *shared_module =
+        (mca_coll_shared_module_t *) module;
+    int size = ompi_comm_size(comm);
+    int rank = ompi_comm_rank(comm);
+    int i;
+    ptrdiff_t extent, lower_bound;
+    ompi_datatype_get_extent(dtype, &lower_bound, &extent);
+
+    /* Enable shared module if necessary */
+    if (!shared_module->enabled) {
+        mca_coll_shared_lazy_enable(module, comm, count * extent);
+    } else if ((size_t) count * extent > shared_module->data_buf_size) {
+        mca_coll_shared_attach_data_buf(shared_module, comm,
+                                        count * extent);
+    }
+    /* Set up segment count */
+    int seg_count, l_seg_count;
+    seg_count = count / size;
+    l_seg_count = seg_count;
+    if (rank == size - 1) {
+        seg_count = count - rank * l_seg_count;
+    }
+    shared_module->ctrl_buf[rank][0] = rank;
+    shared_module->sm_ctrl_win->w_osc_module->osc_fence(0,
```

Comment:
Without providing an implementation for a barrier the cost of the osc_fence (which is a barrier behind the scene) will be gigantic.

Reply:
We should be able to replace the osc_fence with a counted atomic variable.